transports/janus_pfunix.c

```diff
@@ -0,0 +1,526 @@
+#include "../utils.h"
+
+
+/* Transport plugin information */
+#define JANUS_PFUNIX_VERSION			1
+#define JANUS_PFUNIX_VERSION_STRING		"0.0.1"
+#define JANUS_PFUNIX_DESCRIPTION		"This transport plugin adds Unix Sockets support to the Janus API."
+#define JANUS_PFUNIX_NAME				"JANUS Unix Sockets transport plugin"
+#define JANUS_PFUNIX_AUTHOR				"Meetecho s.r.l."
+#define JANUS_PFUNIX_PACKAGE			"janus.transport.pfunix"
+
+/* Transport methods */
+janus_transport *create(void);
+int janus_pfunix_init(janus_transport_callbacks *callback, const char *config_path);
+void janus_pfunix_destroy(void);
+int janus_pfunix_get_api_compatibility(void);
+int janus_pfunix_get_version(void);
+const char *janus_pfunix_get_version_string(void);
+const char *janus_pfunix_get_description(void);
+const char *janus_pfunix_get_name(void);
+const char *janus_pfunix_get_author(void);
+const char *janus_pfunix_get_package(void);
+gboolean janus_pfunix_is_janus_api_enabled(void);
+gboolean janus_pfunix_is_admin_api_enabled(void);
+int janus_pfunix_send_message(void *transport, void *request_id, gboolean admin, json_t *message);
+void janus_pfunix_session_created(void *transport, guint64 session_id);
+void janus_pfunix_session_over(void *transport, guint64 session_id, gboolean timeout);
+
+
+/* Transport setup */
+static janus_transport janus_pfunix_transport =
+	JANUS_TRANSPORT_INIT (
+		.init = janus_pfunix_init,
+		.destroy = janus_pfunix_destroy,
+
+		.get_api_compatibility = janus_pfunix_get_api_compatibility,
+		.get_version = janus_pfunix_get_version,
+		.get_version_string = janus_pfunix_get_version_string,
+		.get_description = janus_pfunix_get_description,
+		.get_name = janus_pfunix_get_name,
+		.get_author = janus_pfunix_get_author,
+		.get_package = janus_pfunix_get_package,
+
+		.is_janus_api_enabled = janus_pfunix_is_janus_api_enabled,
+		.is_admin_api_enabled = janus_pfunix_is_admin_api_enabled,
+
+		.send_message = janus_pfunix_send_message,
+		.session_created = janus_pfunix_session_created,
+		.session_over = janus_pfunix_session_over,
+	);
+
+/* Transport creator */
+janus_transport *create(void) {
+	JANUS_LOG(LOG_VERB, "%s created!\n", JANUS_PFUNIX_NAME);
+	return &janus_pfunix_transport;
+}
+
+
+/* Useful stuff */
+static gint initialized = 0, stopping = 0;
+static janus_transport_callbacks *gateway = NULL;
+
+#define BUFFER_SIZE		8192
+
+
+/* Unix Sockets server thread */
+static GThread *pfunix_thread = NULL;
+void *janus_pfunix_thread(void *data);
+
+/* Unix Sockets servers */
+static int pfd = -1, admin_pfd = -1;
+/* Socket pair to notify about the need for outgoing data */
+static int write_fd[2];
+
+/* Unix Sockets client session */
+typedef struct janus_pfunix_client {
+	int fd;					/* Client socket */
+	gboolean admin;			/* Whether this client is for the Admin or Janus API */
+	GAsyncQueue *messages;	/* Queue of outgoing messages to push */
+	gint session_timeout:1;	/* Whether a Janus session timeout occurred in the core */
```

Comment:
I think I originally meant this to become an atomic int, but boolean sounds more apt as it's all basically single threaded anyway.